lsp.go

```diff
@@ -17,83 +17,132 @@ package tmlsp
-	switch r.Method() {
-	default:
-		log.Printf("%s is not implemented", r.Method())
-	case lsp.MethodInitialize:
-		var params initParams
-		if err := json.Unmarshal(r.Params(), &params); err != nil {
-			log.Fatal(err)
-		}
-
-		s.workspace = string(uri.New(params.RootURI).Filename())
-		err := reply(ctx, lsp.InitializeResult{
-			Capabilities: lsp.ServerCapabilities{
-				CompletionProvider: &lsp.CompletionOptions{},
-
-				// if we support `goto` definition.
-				DefinitionProvider: false,
-
-				// If we support `hover` info.
-				HoverProvider: false,
-
-				TextDocumentSync: lsp.TextDocumentSyncOptions{
-					// Send all file content on every change (can be optimized later).
-					Change: lsp.TextDocumentSyncKindFull,
-
-					// if we want to be notified about open/close of Terramate files.
-					OpenClose: true,
-					Save: &lsp.SaveOptions{
-						// If we want the file content on save,
-						IncludeText: false,
-					},
-				},
-			},
-		}, nil)
+	logger := s.log.With().
+		Str("action", "server.Handler()").
+		Str("method", r.Method()).
+		Logger()
+
+	logger.Debug().
+		RawJSON("params", r.Params()).
+		Msg("handling request.")
+
+	if handler, ok := s.handlers[r.Method()]; ok {
+		return handler(logger, ctx, reply, r)
+	}
+
+	logger.Trace().Msg("not implemented")
+	return nil
+}
+
+func (s *Server) handleInitialize(
+	log zerolog.Logger,
+	ctx context.Context,
+	reply jsonrpc2.Replier,
+	r jsonrpc2.Request,
+) error {
+	type initParams struct {
+		ProcessID int    `json:"processId,omitempty"`
+		RootURI   string `json:"rootUri,omitempty"`
+	}
+
+	var params initParams
+	if err := json.Unmarshal(r.Params(), &params); err != nil {
+		log.Fatal().Err(err).Msg("failed to unmarshal params")
+	}
+
+	s.workspace = string(uri.New(params.RootURI).Filename())
+	err := reply(ctx, lsp.InitializeResult{
+		Capabilities: lsp.ServerCapabilities{
+			CompletionProvider: &lsp.CompletionOptions{},
 
-		if err != nil {
-			log.Fatal(err)
-		}
+			// if we support `goto` definition.
+			DefinitionProvider: false,
 
-		log.Printf("client connected using workspace %q", s.workspace)
+			// If we support `hover` info.
+			HoverProvider: false,
 
-		err = s.conn.Notify(ctx, lsp.MethodWindowShowMessage, lsp.ShowMessageParams{
-			Message: "connected to terramate-lsp",
-			Type:    lsp.MessageTypeInfo,
-		})
+			TextDocumentSync: lsp.TextDocumentSyncOptions{
+				// Send all file content on every change (can be optimized later).
+				Change: lsp.TextDocumentSyncKindFull,
+
+				// if we want to be notified about open/close of Terramate files.
+				OpenClose: true,
+				Save: &lsp.SaveOptions{
+					// If we want the file content on save,
+					IncludeText: false,
+				},
+			},
+		},
+	}, nil)
+
+	if err != nil {
+		log.Fatal().Err(err).Msg("failed to reply")
+	}
 
-		if err != nil {
-			log.Printf("error: failed to notify client")
-		}
+	log.Info().Msgf("client connected using workspace %q", s.workspace)
```

Comment:
```suggestion
	log.Info().
		Str("workspace", s.workspace).
		Msg("client connected")
```

Or maybe even re-create the log instance now with the workspace field, seems like it is useful info.